common/firestore/FirestoreProtoClient.java

```diff
@@ -0,0 +1,270 @@
+import com.google.startupos.common.firestore.Protos.ProtoDocument;
+import com.google.api.core.ApiFuture;
+import com.google.cloud.firestore.WriteResult;
+import java.util.concurrent.ExecutionException;
+import com.google.cloud.firestore.DocumentSnapshot;
+import com.google.cloud.firestore.EventListener;
+import com.google.cloud.firestore.FirestoreException;
+import javax.annotation.Nullable;
+import java.util.Date;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableList;
+import com.google.protobuf.Parser;
+import com.google.protobuf.InvalidProtocolBufferException;
+
+import java.io.InputStream;
+import java.io.FileInputStream;
+import java.net.HttpURLConnection;
+import java.net.URL;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.HashMap;
+import java.io.IOException;
+
+public class FirestoreProtoClient {
+  private static final String PROTO_FIELD = "proto";
+
+  Firestore client;
+
+  public FirestoreProtoClient(String serviceAccountJson) {
+    try {
+      InputStream serviceAccount = new FileInputStream(serviceAccountJson);
+      GoogleCredentials credentials = GoogleCredentials.fromStream(serviceAccount);
+      FirebaseOptions options = new FirebaseOptions.Builder().setCredentials(credentials).build();
+      FirebaseApp.initializeApp(options);
+      client = FirestoreClient.getFirestore();
+    } catch (IOException e) {
+      e.printStackTrace();
+    }
+  }
+
+  public FirestoreProtoClient(String project, String token) {
+    FirebaseOptions options =
+        new FirebaseOptions.Builder()
+            .setCredentials(GoogleCredentials.of(new AccessToken(token, null)))
+            .setProjectId(project)
+            .build();
+    FirebaseApp.initializeApp(options);
+    client = FirestoreClient.getFirestore();
+  }
+
+  public Firestore getClient() {
+    return client;
+  }
+
+  private String joinPath(String collection, String documentId) {
+    if (collection.endsWith("/")) {
+      return collection + documentId;
+    }
+    return collection + "/" + documentId;
+  }
+
+  private Message parseProto(DocumentSnapshot document, Message.Builder builder)
+      throws InvalidProtocolBufferException {
+    return builder
+        .build()
+        .getParserForType()
+        .parseFrom(Base64.getDecoder().decode(document.getString(PROTO_FIELD)));
+  }
+
+  private CollectionReference getCollectionReference(String[] parts, int length) {
+    DocumentReference docRef;
+    CollectionReference collectionRef = client.collection(parts[0]);
+    for (int i = 1; i < length; i += 2) {
+      docRef = collectionRef.document(parts[i]);
+      collectionRef = docRef.collection(parts[i + 1]);
+    }
+    return collectionRef;
+  }
+
+  public DocumentReference getDocumentReference(String path) {
+    String[] parts = path.split("/");
+    if (parts.length % 2 != 0) {
+      throw new IllegalArgumentException("path length should be even but is " + parts.length);
```

Comment:
Maybe `s / path / Path` ?
Also in 209 line.

Reply:
Done